asteroidsyncserviced/openweathermapparser.cpp

```diff
@@ -0,0 +1,101 @@
+#include "openweathermapparser.h"
+
+OpenWeatherMapParser::OpenWeatherMapParser(QObject *parent) : QObject(parent)
+{
+}
+
+void OpenWeatherMapParser::prepareData(QJsonObject obj)
+{
+    m_weatherId.clear();
+    m_tempMin.clear();
+    m_tempMax.clear();
+    m_dateTempMin.setDate(0, 0, 0);
+    m_dateTempMax.setDate(0, 0, 0);
+
+    QDateTime timestamp;
+    QJsonArray list = obj.value("list").toArray();
+
+    for(auto element : list) {
+        QJsonObject dayInfo = element.toObject();
+        timestamp.setTime_t(dayInfo.value("dt").toInt());
+
+        setWeatherId(dayInfo.value("weather").toArray(), timestamp.time());
```

Comment:
I see that every method has a different way of determining "what day it is" but ultimately it should be the same day for every method.
- setWeatherId relies on the assumption that all previous days had a weatherId somewhere between 11 and 13. What if the measurements contain a value at 11:00:00 and 13:00:00 ? You would suddenly inject 2 values in your array and then every following day would be shifted.
- setMaxTemp/setMinTemp both do the same calculation keeping "m_dateTempMin" and "m_dateTempMax" which are the actually the same. Here too, you're making many assumptions, for example about the order in which events are received, if two events are swapped (maybe by the parser or for some other random reason?) your logic becomes broken. You use this to access "m_tempMax.size() - 1" but this creates special cases for when the array is empty etc... By reducing your number of assumptions you can make the code much more simple and generic.

I propose that you extract here in this loop once for each element the "day index" which is going to be the index at which any value will be inserted (whether that is a weather id or a max/min temperature in their respective arrays). You could calculate this offset quite easily with https://doc.qt.io/qt-5/qdate.html#daysTo by comparing the date to the first date in the json document. If you provide this day index to all your methods, you should find that your code becomes greatly simplified. Especially your "setM**Temp" functions should have less special cases for empty arrays/new dates etc... They could just insert their value at the right index. This also provides you stronger guarantees that the "weatherId"/"maxTemp"/"minTemp" will be aligned.